File: draw.py
from display import *

def draw_line( screen, x0, y0, x1, y1, color ):
# *_Algorithm_* (Octant 2)
#   (x_o, y_o) -> (x_1, y_1)
#   x=x_o, y=y_o
#   A = y_1 - y_o, B = -(x_1 - x_o)
#   d = A + 2B
#   while(y \leq y_1)
#     plot (x, y)
#     if (d<0)
#       x+=1
#       d+=2A
#     y+=1
#     d+=2B
#
# draw_line( screen, 0, 0, XRES - 75, YRES - 1, color )
    #Octant 2
    if ((x1>x0)==False):
        return
    if (((y1-y0)/(x1-x0)>1)==False):
        return
    x = x0
    y = y0
    A = y1-y0
    B = -(x1-x0)
    d = A + 2*B
    while(y <= y1):
        plot( screen, color, x, y )
        if (d<0):
            x+=1
            d+=2*A
        y+=1
        d+=2*B
    print("inside draw_line")
    return

File: draw.c
#include <stdio.h>
#include <stdlib.h>

#include "ml6.h"
#include "display.h"
#include "draw.h"

//Insert your line algorithm here
void draw_line(int x0, int y0, int x1, int y1, screen s, color c) {

  int x, y, A, B, d;
  float m;

  //Switch coors
  if (!(x1>x0)){
    int tx, ty;
    tx = x0; ty = y0;
    x0 = x1; y0 = y1;
    x1 = tx; y1 = ty;
  }

  //Vertical line, octants II and VII
  if ((x1-x0)==0){
    x = x0; y = y0;
    A = y1-y0; B = -(x1-x0);
    d = A+2*B;
    while(y <= y1){
      plot(s,c,x,y);
      y+=1;
      d+=2*B;
    }
    x = x0; y = y0;
    A = y1-y0; B = -(x1-x0);
    d = A-2*B;
    while(y >= y1){
      plot(s,c,x,y);
      y-=1;
      d-=2*B;
    }
    return;
  }

  m = ((float)(y1-y0))/((float)(x1-x0));

  //OCTANT I, V
  if ((m <= 1 && m >= 0)){
    x = x0; y = y0;
    A = y1-y0; B = -(x1-x0);
    d = 2*A + B;
    while(x <= x1){
      plot(s, c, x, y);
      if (d>0){
	y+=1;
	d+=2*B;
      }
      x+=1;
      d+=2*A;
    }
  }

  //OCTANT II, VI
  if ((m >= 1)){
    x = x0; y = y0;
    A = y1-y0; B = -(x1-x0);
    d = A+2*B;
    while(y <= y1){
      plot(s,c,x,y);
      if (d<0){
	x+=1;
	d+=2*A;
      }
      y+=1;
      d+=2*B;
    }
  }

  //OCTANT VIII, IV
  if ((m >= -1 && m <= 0 )){
    x = x0; y = y0;
    A = y1-y0; B = -(x1-x0);
    d = 2*A-B;
    while(x <= x1){
      plot(s,c,x,y);
      if (d<0){
	y-=1;
	d-=2*B;
      }
      x+=1;
      d+=2*A;
    }
  }

  //OCTANT VII, III
  if ((m <= -1 )){
    x = x0; y = y0;
    A = y1-y0; B = -(x1-x0);
    d = A-2*B;
    while(y >= y1){
      plot(s,c,x,y);
      if (d>0){
	x+=1;
	d+=2*A;
      }
      y-=1;
      d-=2*B;
    }
  }

}
